### PR TITLE
machine/usb/hid: add MediaKey support

### PR DESCRIPTION
**Pull request by sago35**
#3171 

This PR adds mediakey support.
It was confirmed to work properly in the following environment.

* rp2040 (feather-rp2040)
* samd51 (wioterminal)
* samd21 (feather-m0)
* nrf52840 (feather-nrf52840-sense)

The following source code has not been added to `src/examples/hid-keyboard-mediakey`.
Add if necessary.

```go
package main

import (
	"machine"
	"machine/usb/hid/keyboard"
	"time"
)

func main() {
	button := machine.D5 // BUTTON
	button.Configure(machine.PinConfig{Mode: machine.PinInputPullup})

	kb := keyboard.Port()

	state := 0
	for {
		switch state % 4 {
		case 0:
			if !button.Get() {
				kb.Down(keyboard.KeyMediaVolumeInc)
				state++
			}
		case 1:
			if button.Get() {
				kb.Up(keyboard.KeyMediaVolumeInc)
				state++
			}
		case 2:
			if !button.Get() {
				kb.Down(keyboard.KeyMediaVolumeDec)
				state++
			}
		case 3:
			if button.Get() {
				kb.Up(keyboard.KeyMediaVolumeDec)
				state++
			}
		}
		time.Sleep(20 * time.Millisecond)
	}
}
```

### CODE REVIEW

**Comment by sago35** [comment]
`System Key` is not supported at this time.

```go
	// KeySystemXXX is not supported now
	KeySystemPowerDown Keycode = 0x81 | 0xE200
	KeySystemSleep     Keycode = 0x82 | 0xE200
	KeySystemWakeUp    Keycode = 0x83 | 0xE200
```

**Comment by bgould** [comment]
🔥🔥🔥

this is awesome, going to test right away

**Comment by bgould** [comment]
LGTM ... tested example code on nrf52840 and is working great.  Should be ready to merge after the 0.27 release.  Thank you @sago35 !

**Comment by neildavis** [comment]
@sago35 Thanks for working on this. Unfortunately this did not work for me. Only 'Volume Down' works. 'Volume Inc' and 'Mute' seem to be reported by udev as `code 240: KEY_UNKNOWN`

My setup:

Linux Mint 21.1
TinyGo amd64 `.deb` installed from [the build run linked here](https://github.com/tinygo-org/tinygo/actions/runs/4133199983/)
Pimoroni Tiny2040 2MB (`-target=pico`)

Test program:
```go
package main

import (
	"machine"
	"machine/usb/hid/keyboard"
	"time"
)

func main() {
	buttonCfg := machine.PinConfig{Mode: machine.PinInputPullup}
	buttonInc := machine.GPIO7
	buttonInc.Configure(buttonCfg)
	buttonDec := machine.GPIO6
	buttonDec.Configure(buttonCfg)
	buttonMute := machine.GPIO3
	buttonMute.Configure(buttonCfg)

	kb := keyboard.Port()

	for {
		if !buttonInc.Get() {
			println("DOWN: Vol Inc")
			kb.Down(keyboard.KeyMediaVolumeInc)
		} else if buttonInc.Get() {
			println("UP: Vol Inc")
			kb.Up(keyboard.KeyMediaVolumeInc)
		}
		if !buttonDec.Get() {
			println("DOWN: Vol Dec")
			kb.Down(keyboard.KeyMediaVolumeDec)
		} else if buttonDec.Get() {
			println("UP: Vol Dec")
			kb.Up(keyboard.KeyMediaVolumeDec)
		}
		if !buttonMute.Get() {
			println("DOWN: Mute")
			kb.Down(keyboard.KeyMediaMute)
		} else if buttonMute.Get() {
			kb.Up(keyboard.KeyMediaMute)
			println("Up: Mute")
		}
		time.Sleep(20 * time.Millisecond)
	}
}
```
Input devices as seen by Linux:
```
$ ls -al /dev/input/by-id/
total 0
drwxr-xr-x 2 root root 200 Feb 13 19:51 .
drwxr-xr-x 4 root root 600 Feb 13 19:51 ..
lrwxrwxrwx 1 root root   9 Feb 13 18:58 usb-Logitech_G403_HERO_Gaming_Mouse_029239633931-event-mouse -> ../event7
lrwxrwxrwx 1 root root   9 Feb 13 18:58 usb-Logitech_G403_HERO_Gaming_Mouse_029239633931-if01-event-kbd -> ../event8
lrwxrwxrwx 1 root root   9 Feb 13 18:58 usb-Logitech_G403_HERO_Gaming_Mouse_029239633931-mouse -> ../mouse0
lrwxrwxrwx 1 root root   9 Feb 13 19:51 usb-Raspberry_Pi_Pico-if02-event-kbd -> ../event9
lrwxrwxrwx 1 root root  10 Feb 13 19:51 usb-Raspberry_Pi_Pico-if02-event-mouse -> ../event10
lrwxrwxrwx 1 root root   9 Feb 13 19:51 usb-Raspberry_Pi_Pico-if02-mouse -> ../mouse1
lrwxrwxrwx 1 root root   9 Feb 13 18:58 usb-SONiX_USB_DEVICE-event-if01 -> ../event4
lrwxrwxrwx 1 root root   9 Feb 13 18:58 usb-SONiX_USB_DEVICE-event-kbd -> ../event3
```
EVTEST output for buttons pressed in this order:
1. Dec
2. Inc
3. Mute
```
$ evtest /dev/input/event9
Input driver version is 1.0.1
Input device ID: bus 0x3 vendor 0x2e8a product 0xa version 0x101
Input device name: "Raspberry Pi Pico Keyboard"
Supported events:
  Event type 0 (EV_SYN)
  Event type 1 (EV_KEY)
    Event code 1 (KEY_ESC)
    Event code 2 (KEY_1)
    Event code 3 (KEY_2)
    Event code 4 (KEY_3)
    Event code 5 (KEY_4)
    Event code 6 (KEY_5)
    Event code 7 (KEY_6)
    Event code 8 (KEY_7)
    Event code 9 (KEY_8)
    Event code 10 (KEY_9)
    Event code 11 (KEY_0)
    Event code 12 (KEY_MINUS)
    Event code 13 (KEY_EQUAL)
    Event code 14 (KEY_BACKSPACE)
    Event code 15 (KEY_TAB)
    Event code 16 (KEY_Q)
    Event code 17 (KEY_W)
    Event code 18 (KEY_E)
    Event code 19 (KEY_R)
    Event code 20 (KEY_T)
    Event code 21 (KEY_Y)
    Event code 22 (KEY_U)
    Event code 23 (KEY_I)
    Event code 24 (KEY_O)
    Event code 25 (KEY_P)
    Event code 26 (KEY_LEFTBRACE)
    Event code 27 (KEY_RIGHTBRACE)
    Event code 28 (KEY_ENTER)
    Event code 29 (KEY_LEFTCTRL)
    Event code 30 (KEY_A)
    Event code 31 (KEY_S)
    Event code 32 (KEY_D)
    Event code 33 (KEY_F)
    Event code 34 (KEY_G)
    Event code 35 (KEY_H)
    Event code 36 (KEY_J)
    Event code 37 (KEY_K)
    Event code 38 (KEY_L)
    Event code 39 (KEY_SEMICOLON)
    Event code 40 (KEY_APOSTROPHE)
    Event code 41 (KEY_GRAVE)
    Event code 42 (KEY_LEFTSHIFT)
    Event code 43 (KEY_BACKSLASH)
    Event code 44 (KEY_Z)
    Event code 45 (KEY_X)
    Event code 46 (KEY_C)
    Event code 47 (KEY_V)
    Event code 48 (KEY_B)
    Event code 49 (KEY_N)
    Event code 50 (KEY_M)
    Event code 51 (KEY_COMMA)
    Event code 52 (KEY_DOT)
    Event code 53 (KEY_SLASH)
    Event code 54 (KEY_RIGHTSHIFT)
    Event code 55 (KEY_KPASTERISK)
    Event code 56 (KEY_LEFTALT)
    Event code 57 (KEY_SPACE)
    Event code 58 (KEY_CAPSLOCK)
    Event code 59 (KEY_F1)
    Event code 60 (KEY_F2)
    Event code 61 (KEY_F3)
    Event code 62 (KEY_F4)
    Event code 63 (KEY_F5)
    Event code 64 (KEY_F6)
    Event code 65 (KEY_F7)
    Event code 66 (KEY_F8)
    Event code 67 (KEY_F9)
    Event code 68 (KEY_F10)
    Event code 69 (KEY_NUMLOCK)
    Event code 70 (KEY_SCROLLLOCK)
    Event code 71 (KEY_KP7)
    Event code 72 (KEY_KP8)
    Event code 73 (KEY_KP9)
    Event code 74 (KEY_KPMINUS)
    Event code 75 (KEY_KP4)
    Event code 76 (KEY_KP5)
    Event code 77 (KEY_KP6)
    Event code 78 (KEY_KPPLUS)
    Event code 79 (KEY_KP1)
    Event code 80 (KEY_KP2)
    Event code 81 (KEY_KP3)
    Event code 82 (KEY_KP0)
    Event code 83 (KEY_KPDOT)
    Event code 85 (KEY_ZENKAKUHANKAKU)
    Event code 86 (KEY_102ND)
    Event code 87 (KEY_F11)
    Event code 88 (KEY_F12)
    Event code 89 (KEY_RO)
    Event code 90 (KEY_KATAKANA)
    Event code 91 (KEY_HIRAGANA)
    Event code 92 (KEY_HENKAN)
    Event code 93 (KEY_KATAKANAHIRAGANA)
    Event code 94 (KEY_MUHENKAN)
    Event code 95 (KEY_KPJPCOMMA)
    Event code 96 (KEY_KPENTER)
    Event code 97 (KEY_RIGHTCTRL)
    Event code 98 (KEY_KPSLASH)
    Event code 99 (KEY_SYSRQ)
    Event code 100 (KEY_RIGHTALT)
    Event code 102 (KEY_HOME)
    Event code 103 (KEY_UP)
    Event code 104 (KEY_PAGEUP)
    Event code 105 (KEY_LEFT)
    Event code 106 (KEY_RIGHT)
    Event code 107 (KEY_END)
    Event code 108 (KEY_DOWN)
    Event code 109 (KEY_PAGEDOWN)
    Event code 110 (KEY_INSERT)
    Event code 111 (KEY_DELETE)
    Event code 113 (KEY_MUTE)
    Event code 114 (KEY_VOLUMEDOWN)
    Event code 115 (KEY_VOLUMEUP)
    Event code 116 (KEY_POWER)
    Event code 117 (KEY_KPEQUAL)
    Event code 119 (KEY_PAUSE)
    Event code 121 (KEY_KPCOMMA)
    Event code 122 (KEY_HANGUEL)
    Event code 123 (KEY_HANJA)
    Event code 124 (KEY_YEN)
    Event code 125 (KEY_LEFTMETA)
    Event code 126 (KEY_RIGHTMETA)
    Event code 127 (KEY_COMPOSE)
    Event code 128 (KEY_STOP)
    Event code 129 (KEY_AGAIN)
    Event code 130 (KEY_PROPS)
    Event code 131 (KEY_UNDO)
    Event code 132 (KEY_FRONT)
    Event code 133 (KEY_COPY)
    Event code 134 (KEY_OPEN)
    Event code 135 (KEY_PASTE)
    Event code 136 (KEY_FIND)
    Event code 137 (KEY_CUT)
    Event code 138 (KEY_HELP)
    Event code 139 (KEY_MENU)
    Event code 140 (KEY_CALC)
    Event code 142 (KEY_SLEEP)
    Event code 144 (KEY_FILE)
    Event code 150 (KEY_WWW)
    Event code 152 (KEY_SCREENLOCK)
    Event code 155 (KEY_MAIL)
    Event code 156 (KEY_BOOKMARKS)
    Event code 158 (KEY_BACK)
    Event code 159 (KEY_FORWARD)
    Event code 161 (KEY_EJECTCD)
    Event code 163 (KEY_NEXTSONG)
    Event code 164 (KEY_PLAYPAUSE)
    Event code 165 (KEY_PREVIOUSSONG)
    Event code 166 (KEY_STOPCD)
    Event code 167 (KEY_RECORD)
    Event code 168 (KEY_REWIND)
    Event code 169 (KEY_PHONE)
    Event code 171 (KEY_CONFIG)
    Event code 172 (KEY_HOMEPAGE)
    Event code 173 (KEY_REFRESH)
    Event code 174 (KEY_EXIT)
    Event code 176 (KEY_EDIT)
    Event code 177 (KEY_SCROLLUP)
    Event code 178 (KEY_SCROLLDOWN)
    Event code 179 (KEY_KPLEFTPAREN)
    Event code 180 (KEY_KPRIGHTPAREN)
    Event code 181 (KEY_NEW)
    Event code 182 (KEY_REDO)
    Event code 183 (KEY_F13)
    Event code 184 (KEY_F14)
    Event code 185 (KEY_F15)
    Event code 186 (KEY_F16)
    Event code 187 (KEY_F17)
    Event code 188 (KEY_F18)
    Event code 189 (KEY_F19)
    Event code 190 (KEY_F20)
    Event code 191 (KEY_F21)
    Event code 192 (KEY_F22)
    Event code 193 (KEY_F23)
    Event code 194 (KEY_F24)
    Event code 206 (KEY_CLOSE)
    Event code 207 (KEY_PLAY)
    Event code 208 (KEY_FASTFORWARD)
    Event code 209 (KEY_BASSBOOST)
    Event code 210 (KEY_PRINT)
    Event code 212 (KEY_CAMERA)
    Event code 216 (KEY_CHAT)
    Event code 217 (KEY_SEARCH)
    Event code 219 (KEY_FINANCE)
    Event code 223 (KEY_CANCEL)
    Event code 224 (KEY_BRIGHTNESSDOWN)
    Event code 225 (KEY_BRIGHTNESSUP)
    Event code 228 (KEY_KBDILLUMTOGGLE)
    Event code 229 (KEY_KBDILLUMDOWN)
    Event code 230 (KEY_KBDILLUMUP)
    Event code 231 (KEY_SEND)
    Event code 232 (KEY_REPLY)
    Event code 233 (KEY_FORWARDMAIL)
    Event code 234 (KEY_SAVE)
    Event code 235 (KEY_DOCUMENTS)
    Event code 240 (KEY_UNKNOWN)
    Event code 241 (KEY_VIDEO_NEXT)
    Event code 244 (KEY_BRIGHTNESS_ZERO)
    Event code 256 (BTN_0)
    Event code 353 (KEY_SELECT)
    Event code 354 (KEY_GOTO)
    Event code 358 (KEY_INFO)
    Event code 362 (KEY_PROGRAM)
    Event code 366 (KEY_PVR)
    Event code 370 (KEY_SUBTITLE)
    Event code 372 (KEY_ZOOM)
    Event code 374 (KEY_KEYBOARD)
    Event code 375 (KEY_SCREEN)
    Event code 376 (KEY_PC)
    Event code 377 (KEY_TV)
    Event code 378 (KEY_TV2)
    Event code 379 (KEY_VCR)
    Event code 380 (KEY_VCR2)
    Event code 381 (KEY_SAT)
    Event code 383 (KEY_CD)
    Event code 384 (KEY_TAPE)
    Event code 386 (KEY_TUNER)
    Event code 387 (KEY_PLAYER)
    Event code 389 (KEY_DVD)
    Event code 392 (KEY_AUDIO)
    Event code 393 (KEY_VIDEO)
    Event code 396 (KEY_MEMO)
    Event code 397 (KEY_CALENDAR)
    Event code 398 (KEY_RED)
    Event code 399 (KEY_GREEN)
    Event code 400 (KEY_YELLOW)
    Event code 401 (KEY_BLUE)
    Event code 402 (KEY_CHANNELUP)
    Event code 403 (KEY_CHANNELDOWN)
    Event code 405 (KEY_LAST)
    Event code 407 (KEY_NEXT)
    Event code 408 (KEY_RESTART)
    Event code 409 (KEY_SLOW)
    Event code 410 (KEY_SHUFFLE)
    Event code 412 (KEY_PREVIOUS)
    Event code 416 (KEY_VIDEOPHONE)
    Event code 417 (KEY_GAMES)
    Event code 418 (KEY_ZOOMIN)
    Event code 419 (KEY_ZOOMOUT)
    Event code 420 (KEY_ZOOMRESET)
    Event code 421 (KEY_WORDPROCESSOR)
    Event code 422 (KEY_EDITOR)
    Event code 423 (KEY_SPREADSHEET)
    Event code 424 (KEY_GRAPHICSEDITOR)
    Event code 425 (KEY_PRESENTATION)
    Event code 426 (KEY_DATABASE)
    Event code 427 (KEY_NEWS)
    Event code 428 (KEY_VOICEMAIL)
    Event code 429 (KEY_ADDRESSBOOK)
    Event code 430 (KEY_MESSENGER)
    Event code 431 (KEY_DISPLAYTOGGLE)
    Event code 432 (KEY_SPELLCHECK)
    Event code 433 (KEY_LOGOFF)
    Event code 439 (KEY_MEDIA_REPEAT)
    Event code 442 (KEY_IMAGES)
    Event code 576 (KEY_BUTTONCONFIG)
    Event code 577 (KEY_TASKMANAGER)
    Event code 578 (KEY_JOURNAL)
    Event code 579 (KEY_CONTROLPANEL)
    Event code 580 (KEY_APPSELECT)
    Event code 581 (KEY_SCREENSAVER)
    Event code 582 (KEY_VOICECOMMAND)
    Event code 583 (KEY_ASSISTANT)
    Event code 585 (?)
    Event code 586 (?)
    Event code 592 (KEY_BRIGHTNESS_MIN)
    Event code 593 (KEY_BRIGHTNESS_MAX)
  Event type 2 (EV_REL)
    Event code 6 (REL_HWHEEL)
    Event code 12 (REL_HWHEEL_HI_RES)
  Event type 3 (EV_ABS)
    Event code 32 (ABS_VOLUME)
      Value      0
      Min        0
      Max      652
  Event type 4 (EV_MSC)
    Event code 4 (MSC_SCAN)
Key repeat handling:
  Repeat type 20 (EV_REP)
    Repeat code 0 (REP_DELAY)
      Value    250
    Repeat code 1 (REP_PERIOD)
      Value     33
Properties:
Testing ... (interrupt to exit)
Event: time 1676318440.591824, type 4 (EV_MSC), code 4 (MSC_SCAN), value c00ea
Event: time 1676318440.591824, type 1 (EV_KEY), code 114 (KEY_VOLUMEDOWN), value 1
Event: time 1676318440.591824, -------------- SYN_REPORT ------------
Event: time 1676318440.751824, type 4 (EV_MSC), code 4 (MSC_SCAN), value c00ea
Event: time 1676318440.751824, type 1 (EV_KEY), code 114 (KEY_VOLUMEDOWN), value 0
Event: time 1676318440.751824, -------------- SYN_REPORT ------------
Event: time 1676318441.992853, type 4 (EV_MSC), code 4 (MSC_SCAN), value c00eb
Event: time 1676318441.992853, type 1 (EV_KEY), code 240 (KEY_UNKNOWN), value 1
Event: time 1676318441.992853, -------------- SYN_REPORT ------------
Event: time 1676318442.112836, type 4 (EV_MSC), code 4 (MSC_SCAN), value c00eb
Event: time 1676318442.112836, type 1 (EV_KEY), code 240 (KEY_UNKNOWN), value 0
Event: time 1676318442.112836, -------------- SYN_REPORT ------------
Event: time 1676318444.154862, type 4 (EV_MSC), code 4 (MSC_SCAN), value c00e3
Event: time 1676318444.154862, type 1 (EV_KEY), code 240 (KEY_UNKNOWN), value 1
Event: time 1676318444.154862, -------------- SYN_REPORT ------------
Event: time 1676318444.274860, type 4 (EV_MSC), code 4 (MSC_SCAN), value c00e3
Event: time 1676318444.274860, type 1 (EV_KEY), code 240 (KEY_UNKNOWN), value 0
Event: time 1676318444.274860, -------------- SYN_REPORT ------------
```

**Comment by sago35** [comment]
Thanks. I will check. @neildavis

**Comment by sago35** [comment]
USAGE is now specified individually.
Confirmed to work on Windows and Ubuntu22.04.
(The code did not work on ubuntu before the modification.)

I would like someone to confirm this on macOS.
I would also like to see confirmation on Linux Mint 21.1. @neildavis 

https://github.com/ardnew/tinygo/blob/6dcabd49800981e31c94d0ce5a571376c20b854b/src/machine/usb/dhw-hid_atsamd51.go#L126

The code used for the test is as follows.

```go
// tinygo flash --target wioterminal --size short --monitor .

package main

import (
	"machine"
	"machine/usb/hid/keyboard"
	"time"
)

func main() {
	machine.WIO_KEY_A.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	machine.WIO_KEY_B.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	machine.WIO_KEY_C.Configure(machine.PinConfig{Mode: machine.PinInputPullup})

	machine.WIO_5S_UP.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	machine.WIO_5S_LEFT.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	machine.WIO_5S_RIGHT.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	machine.WIO_5S_DOWN.Configure(machine.PinConfig{Mode: machine.PinInputPullup})
	machine.WIO_5S_PRESS.Configure(machine.PinConfig{Mode: machine.PinInputPullup})

	kb := keyboard.Port()

	for {
		if !machine.WIO_KEY_A.Get() {
			kb.Down(keyboard.KeyMediaMute)
		} else {
			kb.Up(keyboard.KeyMediaMute)
		}
		if !machine.WIO_KEY_B.Get() {
			kb.Down(keyboard.KeyMediaFastForward)
		} else {
			kb.Up(keyboard.KeyMediaFastForward)
		}
		if !machine.WIO_KEY_C.Get() {
			kb.Down(keyboard.KeyMediaRewind)
		} else {
			kb.Up(keyboard.KeyMediaRewind)
		}
		if !machine.WIO_5S_UP.Get() {
			kb.Down(keyboard.KeyMediaVolumeInc)
		} else {
			kb.Up(keyboard.KeyMediaVolumeInc)
		}
		if !machine.WIO_5S_LEFT.Get() {
			kb.Down(keyboard.KeyMediaPrevTrack)
		} else {
			kb.Up(keyboard.KeyMediaPrevTrack)
		}
		if !machine.WIO_5S_RIGHT.Get() {
			kb.Down(keyboard.KeyMediaNextTrack)
		} else {
			kb.Up(keyboard.KeyMediaNextTrack)
		}
		if !machine.WIO_5S_DOWN.Get() {
			kb.Down(keyboard.KeyMediaVolumeDec)
		} else {
			kb.Up(keyboard.KeyMediaVolumeDec)
		}
		if !machine.WIO_5S_PRESS.Get() {
			kb.Down(keyboard.KeyMediaPlayPause)
		} else {
			kb.Up(keyboard.KeyMediaPlayPause)
		}
		time.Sleep(50 * time.Millisecond)
	}
}
```

**Comment by neildavis** [comment]
> Confirmed to work on Windows and Ubuntu22.04.
> I would like someone to confirm this on macOS. I would also like to see confirmation on Linux Mint 21.1. @neildavis
> 
@sago35 I can do both and will test this soon. Linux Mint 21 is based on Ubuntu 22.04 but I will test it anyway :)

**Comment by neildavis** [comment]
@sago35 Tested successfully on Linux Mint 21.1 & macOS Monterey

All `KeyMediaXyz` are working except `KeyMediaPlaySkip` which still reports as `code 240 (KEY_UNKNOWN)`

**Comment by sago35** [comment]
@neildavis Thank you.

> All KeyMediaXyz are working except KeyMediaPlaySkip which still reports as code 240 (KEY_UNKNOWN)

I think the definition of playskip is correct and should be unchanged.

https://www.usb.org/sites/default/files/documents/hut1_12v2.pdf

image

----

The way it works now, only defined keys can be handled by TinyGo.
However, I don't see a problem with most use cases.

File: src/machine/usb/descriptor.go [hunk 1]
```diff
@@ -67,7 +67,7 @@ var DescriptorCDCHID = Descriptor{
 		0x07, 0x05, 0x02, 0x02, 0x40, 0x00, 0x00,
 		0x07, 0x05, 0x83, 0x02, 0x40, 0x00, 0x00,
 		0x09, 0x04, 0x02, 0x00, 0x01, 0x03, 0x00, 0x00, 0x00,
-		0x09, 0x21, 0x01, 0x01, 0x00, 0x01, 0x22, 0x65, 0x00,
+		0x09, 0x21, 0x01, 0x01, 0x00, 0x01, 0x22, 0x65 + 27*2 + 1, 0x00,
```
Comment (sago35) [inline src/machine/usb/descriptor.go]: It will be changed to 0x9C later.

0x9C = 0x65 + 27*2 + 1

**Comment by deadprogram** [comment]
So does this mean this PR should be merged, or not?

**Comment by neildavis** [comment]
Just curious, but why is it necessary to define every media key with its own usage in the HID report descriptor? This seems rather verbose and is in contrast to what I see from other peripherals.

As way of an example, using [usbhid-dump](https://manpages.ubuntu.com/manpages/bionic/man8/usbhid-dump.8.html) on Linux and [this HID parsing tool](http://eleccelerator.com/usbdescreqparser/) I dumped my USB keyboard and the RP2040 running TinyGo, and these are the results (limiting output to 'Consumer' usage page `0x0C`):

Keyboard:
```
0x05, 0x0C,        // Usage Page (Consumer)
0x09, 0x01,        // Usage (Consumer Control)
0xA1, 0x01,        // Collection (Application)
0x85, 0x03,        //   Report ID (3)
0x15, 0x00,        //   Logical Minimum (0)
0x26, 0xFF, 0x1F,  //   Logical Maximum (8191)
0x19, 0x00,        //   Usage Minimum (Unassigned)
0x2A, 0xFF, 0x1F,  //   Usage Maximum (0x1FFF)
0x75, 0x10,        //   Report Size (16)
0x95, 0x01,        //   Report Count (1)
0x81, 0x00,        //   Input (Data,Array,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC0,              // End Collection
```

RP2040 & TinyGo:
```
0x05, 0x0C,        // Usage Page (Consumer)
0x09, 0x01,        // Usage (Consumer Control)
0xA1, 0x01,        // Collection (Application)
0x85, 0x03,        //   Report ID (3)
0x05, 0x0C,        //   Usage Page (Consumer)
0x15, 0x00,        //   Logical Minimum (0)
0x25, 0x01,        //   Logical Maximum (1)
0x75, 0x01,        //   Report Size (1)
0x95, 0x0F,        //   Report Count (15)
0x09, 0xB0,        //   Usage (Play)
0x09, 0xB1,        //   Usage (Pause)
0x09, 0xB2,        //   Usage (Record)
0x09, 0xB3,        //   Usage (Fast Forward)
0x09, 0xB4,        //   Usage (Rewind)
0x09, 0xB5,        //   Usage (Scan Next Track)
0x09, 0xB6,        //   Usage (Scan Previous Track)
0x09, 0xB7,        //   Usage (Stop)
0x09, 0xB8,        //   Usage (Eject)
0x09, 0xB9,        //   Usage (Random Play)
0x09, 0xCD,        //   Usage (Play/Pause)
0x09, 0xCE,        //   Usage (Play/Skip)
0x09, 0xE2,        //   Usage (Mute)
0x09, 0xE9,        //   Usage (Volume Increment)
0x09, 0xEA,        //   Usage (Volume Decrement)
0x81, 0x02,        //   Input (Data,Var,Abs,No Wrap,Linear,Preferred State,No Null Position)
0x95, 0x01,        //   Report Count (1)
0x81, 0x01,        //   Input (Const,Array,Abs,No Wrap,Linear,Preferred State,No Null Position)
0xC0,              // End Collection
```

**Comment by sago35** [comment]
It worked on Linux too if I set it to the HID report that @neildavis wrote.
I have checked on Windows and Ubuntu22.04.

**Comment by sago35** [comment]
> So does this mean this PR should be merged, or not?

@deadprogram 
If there is no problem on macOS, it is possible to merge.

**Comment by neildavis** [comment]
Great work @sago35! I will test on macOS later today

**Comment by neildavis** [comment]
@sago35 Works on macOS Monterey. I wasn't able to exhaustively test all of the media keys individually in the time i had, due to lack of a utility like `evtest` on macOS, but the volume inc, dec & mute definitely work ok, and a simple program looping over all of them causes the Music app to launch and start/stop playing, skip tracks etc so it looks like they're all working too! I may have more time to test further on the weekend if required.

**Review by deadprogram — APPROVED** [review]
Looks good to squash/merge. Thanks @sago35 and @bgould + @neildavis for testing/reviewing.

**Pull request merged**